### ISSUE TITLE
Getting 400 on /auth/v1/token?grant_type=refresh_token

### ISSUE THREAD

**Issue by miguelespinoza** [supabase/gotrue-js#454]
# Bug report

## Describe the bug

I'm having an issue where `/auth/v1/token?grant_type=refresh_token` suddenly returns 400. It's unclear what could cause this issue. When the token is approaching expiration supabase-js fires correctly, but it gets to a point where the 400 request occurs. 

This originated after finding out that the supabase token disappears from localStorage on my extension.
The video walks through the situation in more detail: https://share.cleanshot.com/mc6hDzOEoD7lzwk4ZSrk

I've read that supabase v2 fixes "getting logged out” issues, but I'm not so sure this happens from the javascript side. To be precise, I'm initializing the client without the `multiTab` option, so it's not the race condition mentioned throughout the discussions/issues on Github.
Also, the Network logs, are isolated to one process. Meaning only one supabase client. Inside this process, each request uses the previous request's refresh token from the response. Being that it's isolated to only one process, I'm not sure how this could be a library issue. I could be wrong though

This is how I'm creating the supabase client.

```
const browserStorageInterface: SupportedStorage = {
  async getItem(key: string): Promise<string | null> {
    const storage = await browser.storage.local.get(key);
    return storage?.[key];
  },
  setItem(key: string, value: string) {
    browser.storage.local.set({
      [key]: value,
    });
  },
  removeItem(key: string) {
    browser.storage.local.remove(key);
  },
};

const supabase = createClient(supabaseUrl, supabaseKey, {
  autoRefreshToken: true,
  persistSession: true,
  localStorage: browserStorageInterface,
});
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Setup supabase client
2. Authenticate using Google oAuth
3. Set JWT expiry to 20 seconds (to reproduce quickly)
4. supabase refreshes token several times
5. Eventually errors with status code 400.
6. Token is removed from localStorage

## Expected behavior

At no point should refreshing the token return status code 400. 

## Screenshots

https://share.cleanshot.com/mc6hDzOEoD7lzwk4ZSrk

## System information

- OS: macOS
- Browser: chrome - browser extension
- Version of supabase-js: v1.35.3
- Version of Node.js: v12.17.0

## Additional context

Help post on Discord: https://discord.com/channels/839993398554656828/1019328795129421944/1019328795129421944

**Comment by andyjakubowski** [comment]
Thanks for elaborating @miguelespinoza. I’m looking at the `gotrue-js` codebase in the `rc` branch. I presume this is the v2 you are referring to?

The scenario I described makes any and all calls through the Supabase client instance. In my actual code I use the `autoRefreshToken` and let the client instance handle refreshing the access token. I’m merely pointing out that the code which schedules tokens to be refreshed uses `refresh_token` values at the time of scheduling. So if in the meantime anything, for example another tab, actively calls `setSession` for whatever reason, then that scheduled refresh code will be using now-revoked refresh tokens.

I don’t know if this is the cause of the issue you’re experiencing, but I wanted to point out this fact in case it’s relevant.

**Comment by miguelespinoza** [comment]
Yup, we're both on the same page.

That's something that should be accounted for by the library. To pick up the latest `refresh_token`, rather than the value from the time of schedule. FWIW, this is partly working. When making a supabase request in v2, `fetch` verifies if it's using an up to date accessToken. Check out [fetchWithAuth](https://github.com/supabase/supabase-js/blob/v2.0.0-rc.10/src/SupabaseClient.ts#L117). This just has to also apply to the auth logic we're discussing

And yup, you're on the right branch, `rc`.

**Comment by gprieto** [comment]
Hey @miguelespinoza , I have encountered a related issue while trying to set up auth in a browser extension and keeping a session in the popup and service-worker. Would you mind having  a look? Maybe you can shed some light...

https://github.com/supabase/supabase-js/issues/570

**Comment by andyjakubowski** [comment]
> When making a supabase request in v2, `fetch` verifies if it's using an up to date accessToken. Check out [fetchWithAuth](https://github.com/supabase/supabase-js/blob/v2.0.0-rc.10/src/SupabaseClient.ts#L117). This just has to also apply to the auth logic we're discussing

Thanks for pointing this out; I didn’t realize `supabase-js` ran any custom `options.global.fetch` through `fetchWithAuth` first!

> This just has to also apply to the auth logic we're discussing

Yep! 👌

**Comment by kangmingtay** [comment]
Hey everyone, i think i've figured out the issue, seems like the session passed into the `_startAutoRefreshToken` method can possibly become stale and fall outside the reuse interval. I've made a [PR](https://github.com/supabase/gotrue-js/pull/461). Would be awesome if you can try it out and see if it fixes the problem for you!

Tested with the following scenarios:
* JWT_EXPIRY: 30s, REUSE_INTERVAL: 10s, 5 tabs opened (firefox)
* JWT_EXPIRY: 20s, REUSE_INTERVAL: 5s, 5 tabs opened (firefox)

**Comment by kangmingtay** [comment]
just merged #461 in, will bump the supabase-js branch to include this fix next - shoutout to @miguelespinoza, thanks for pointing this out and writing up such a detailed issue!
